src/server/graceful.rs

```diff
@@ -0,0 +1,620 @@
+//! Utility to gracefully shutdown a server.
+//!
+//! This module provides a [`GracefulShutdown`] type,
+//! which can be used to gracefully shutdown a server.
+//!
+//! See <https://github.com/hyperium/hyper-util/blob/master/examples/server_graceful.rs>
+//! for an example of how to use this.
+
+use pin_project_lite::pin_project;
+use slab::Slab;
+use std::{
+    fmt::{self, Debug},
+    future::Future,
+    pin::Pin,
+    sync::{atomic::AtomicUsize, Arc, Mutex},
+    task::{self, Poll, Waker},
+};
+
+/// A graceful shutdown watcher
+pub struct GracefulShutdown {
+    // state used to keep track of all futures that are being watched
+    state: Arc<GracefulState>,
+    // state that the watched futures to know when shutdown signal is received
+    future_state: Arc<GracefulState>,
+}
+
+impl Debug for GracefulShutdown {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("GracefulShutdown").finish()
+    }
+}
+
+impl Default for GracefulShutdown {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GracefulShutdown {
+    /// Create a new graceful shutdown watcher
+    pub fn new() -> Self {
+        Self {
+            state: Arc::new(GracefulState {
+                counter: AtomicUsize::new(0),
+                waker_list: Mutex::new(Slab::new()),
+            }),
+            future_state: Arc::new(GracefulState {
+                counter: AtomicUsize::new(1),
+                waker_list: Mutex::new(Slab::new()),
+            }),
+        }
+    }
+
+    /// Get a graceful shutdown watcher
+    pub fn watcher(&self) -> GracefulWatcher {
+        self.state.subscribe();
+        GracefulWatcher {
+            state: self.state.clone(),
+            future_state: self.future_state.clone(),
+        }
+    }
```

Comment:
We can spawn multiple `GracefulWatcher` from a single `GracefulShutdown`, so is it accurate to say that the states in GracefulShutdown is for every watcher and futures?

Reply:
Yes. That is accurate, they all lead to the same watcher. Once you shutdown however you consume the `GracefulShutdown` and no new graceful watchers can be created.